pkg/slo-controller/noderesource/plugins/gpudeviceresource/plugin.go

```diff
@@ -0,0 +1,223 @@
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/koordinator-sh/koordinator/apis/configuration"
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/framework"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+const PluginName = "GPUDeviceResource"
+
+const (
+	ResetResourcesMsg  = "reset node gpu resources"
+	UpdateResourcesMsg = "node gpu resources from device"
+	UpdateLabelsMsg    = "node gpu labels from device"
+
+	NeedSyncForResourceDiffMsg     = "gpu resource diff is big than threshold"
+	NeedSyncForGPUModelMsg         = "gpu model changed"
+	NeedSyncForGPUDriverVersionMsg = "gpu driver version changed"
+)
+
+var (
+	// ResourceNames are the known resources reconciled by the plugin from devices.
+	// TODO: support custom resources.
+	ResourceNames = []corev1.ResourceName{
+		extension.ResourceGPU,
+		extension.ResourceGPUCore,
+		extension.ResourceGPUMemory,
+		extension.ResourceGPUMemoryRatio,
+	}
+
+	client ctrlclient.Client
+)
+
+type Plugin struct{}
+
+func (p *Plugin) Name() string {
+	return PluginName
+}
+
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;patch
+// +kubebuilder:rbac:groups=core,resources=nodes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups=scheduling.koordinator.sh,resources=devices,verbs=get;list;watch
+// +kubebuilder:rbac:groups=topology.node.k8s.io,resources=noderesourcetopologies,verbs=get;list;watch;create;update
+
+func (p *Plugin) Setup(opt *framework.Option) error {
+	client = opt.Client
+
+	// schedulingv1alpha1.AddToScheme(opt.Scheme)
+	opt.Builder = opt.Builder.Watches(&source.Kind{Type: &schedulingv1alpha1.Device{}}, &DeviceHandler{})
+
+	return nil
+}
+
+func (p *Plugin) NeedSync(strategy *configuration.ColocationStrategy, oldNode, newNode *corev1.Node) (bool, string) {
+	for _, resourceName := range ResourceNames {
+		if util.IsResourceDiff(oldNode.Status.Allocatable, newNode.Status.Allocatable, resourceName,
+			*strategy.ResourceDiffThreshold) {
+			klog.V(4).Infof("node %s resource %s diff bigger than %v, need sync",
+				newNode.Name, resourceName, *strategy.ResourceDiffThreshold)
+			return true, NeedSyncForResourceDiffMsg
+		}
+	}
+
+	return false, ""
+}
+
+func (p *Plugin) NeedSyncMeta(_ *configuration.ColocationStrategy, oldNode, newNode *corev1.Node) (bool, string) {
+	if oldNode.Labels[extension.LabelGPUModel] != newNode.Labels[extension.LabelGPUModel] {
```

Comment:
Making these labels into a table makes it easier to maintain and modify them.

Reply:
fixed